✨ Config v6

randycoulman opened the pull request:
**NOTE:** This PR is based on #116 and the base branch has been set accordingly. I'll rebase before merging.

### Describe the purpose of your pull request

- Feature branch for all config v6 changes

<hr/>

Features/improvements:
* [x] Fetch new config json version (`config_v6.json`) and update config model
* [x] SDK key format validation (when client is not set up to use local-only flag overrides)
* [x] AND conditions
* [x] New condition types
   * Prerequisite flag condition
   * Segment condition
* [x] Changed comparison operators
   * (NOT) CONTAINS ANY OF
* [x] New comparison operators
   * BEFORE/AFTER
   * TEXT (NOT) EQUALS (both hashed and cleartext)
   * TEXT (NOT) STARTS/ENDS WITH ANY OF (both hashed and cleartext)
   * ARRAY CONTAINS ANY OF (both hashed and cleartext)
* [x] Percentage options within targeting rules
* [x] Percentage options based on custom attributes
* [x] Custom User Object attribute handling (don't force string values + autoconversions under the hood)
* [x] User Object attribute/comparison value trimming
* [x] Comparison value hashing changes (SHA256, salting)
* [x] Flag overrides vs. config json salt and segments
* [x] `EvaluationDetails.matched_targeting_rule`/`matched_percentage_option` properties (rename + set combinations correctly)
* [x] Update `ConfigCatClient.getKeyAndValueFromSettingsMap`

Tests: 
* [x] Update cache key/payload tests
* [x] Evaluation matrix tests
  * Config v1 features (both V1 & V2)
  * Config v2 features
* [x] Evaluation log tests
* [x] SDK key format validation
* [x] Custom User Object attribute handling (don't force string values + autoconversions under the hood)
* [x] Flag overrides vs. prerequisite flags (circular dependency, comparison value type mismatch, dependent/prereq flag overriding)
* [x] Flag overrides vs. config json salt and segments
* [x] `EvaluationDetails.matched_targeting_rule`/`matched_percentage_option` properties (rename + set combinations correctly)
* [x] Special character tests (https://trello.com/c/qh4jHrvm)

Related improvements/fixes:
* [x] Dynamically typed languages should log a warning when default value doesn't match the setting type (see https://trello.com/c/Le6vimGu)
* [x] We should go to the cache in all polling modes instead of using the in memory variable (see https://trello.com/c/rreKm64A)
* [x] Improve message of log event 1103 (https://trello.com/c/Mb7rvm64)


### Related issues (only if applicable)

https://trello.com/c/wpNt5Lqu/23-catnip

### Requirement checklist (only if applicable)

- [x] I have covered the applied changes with automated tests.
- [x] I have executed the full automated test set against my changes.
- [x] I have validated my changes against all supported platform versions.
- [x] I have read and accepted the [contribution agreement](https://github.com/configcat/legal/blob/main/contribution-agreement.md).

randycoulman committed:
✨ Update EvaluationFormula for v6

- Add setting_type field
- Rename percentage_rules -> percentage_options
- Rename rollout_rules -> targeting_rules

randycoulman committed:
🚧 Start converting to config v6

Update constants and test literals to new format. Extract sample config to a factory.

randycoulman committed:
✅ Simplify default user tests

No need to sort maps to compare them. Also, add extra keys needed after switching to use the factory config.

randycoulman committed:
🚧 Update config modules for new format

Begin updating the Config sub-modules to support the new Config format.

randycoulman committed:
🚧 Evaluate new targeting rule format

randycoulman committed:
🚧 Extract typed values from percentage options

randycoulman committed:
🚧 Use comparison value of proper type

randycoulman committed:
🚧 Implement segment evaluation

randycoulman committed:
✨ Extract segments from config

randycoulman committed:
✅ Handle string lists in comparators

randycoulman committed:
✨ Use salt for hashing sensitive values

randycoulman committed:
♻️ Refactor EvaluationFormula.variation_value

randycoulman committed:
♻️ Merge Rollout.Comparator into Config.UserComparator

randycoulman committed:
♻️ Delete to ValueAndVariationId where possible

randycoulman committed:
♻️ Use Context struct to hold evaluation context

randycoulman committed:
✅ Update matrix tests to v2 format

Use v2 SDK Keys and add comments with URLs pointing to the relevant dashboard

randycoulman committed:
♻️ Default to value_type_test

randycoulman committed:
🧪 Add (and skip) new matrix tests

These tests require new features that are not yet implemented, so skipping for now until the support has been added.

randycoulman committed:
♻️ Rename comparators and descriptions

Rename a few of the comparators and simplify descriptions to match latest Python SDK.

randycoulman committed:
♻️ Rename EvaluationDetails fields

randycoulman committed:
✨ Support before and after comparators

randycoulman committed:
✨ Support equals/not_equals hashed comparators

Refactor comparator implementations to inline where it makes sense.

randycoulman committed:
✨ Implement starts/ends with any of hashed comparators

randycoulman committed:
🐛 Fix (not_)contains_any_of to work with string lists

Use plural `comparison_values` when a comparator takes a string list to make it clearer.

randycoulman committed:
✨ Implement array (not) contains any of comparators

randycoulman committed:
✨ Implement equals/not equals comparators

randycoulman committed:
✨ Implement (not) starts/ends with comparators

randycoulman committed:
✨ Implement array (not) contains any of comparators

randycoulman committed:
💡 Update skip reasons for matrix tests

randycoulman committed:
✨ Update comment/SDK key for segments_old test to v2

randycoulman committed:
♻️ Push nil user checks down

randycoulman committed:
♻️ Update type conversion code

Ensure values are converted to types consistently, returning errors for invalid conversions.

randycoulman committed:
✨ Support NaiveDateTime by converting to UTC

randycoulman committed:
✅ Parse expected values in matrix tests

This allows us to enable more of the matrix tests.

randycoulman committed:
🐛 Fix handling of Unicode strings

The `length` provided in the expected values is a byte length, not a character length, so parse the strings accordingly.

randycoulman committed:
✅ Add attribute value conversion tests

In order to test non-UTC timezones, we need a timezone database. I've added one for the `test` environment only and I make it the default in `test_helper.exs`.

In order to pass the tests, I had to adjust the implementation of `to_unix_seconds` to return fractional seconds.

Note that Elixir does not support +inf/-inf/NaN float values, so I had to leave those tests commented out.

randycoulman committed:
✨ Support nested percentage options

No tests yet; need custom percentage attribute in order for the supplied tests to pass. That will come in the next commit.

randycoulman committed:
✨ Support custom percentage attribute

randycoulman committed:
🐛 Return immediately on condition evaluation error

randycoulman committed:
♻️ Simplify percentage option evaluation

randycoulman committed:
✨ Support prerequisite flag conditions

Basic support only; still need to check for type mismatches and circular dependencies. Also need to ensure all logging goes to the same log instance.

randycoulman committed:
🐛 Move evaluation log up to caller

We want to use the same logging agent for all recursive calls to `evaluate`, so move the log management up to the caller and pass the log agent's pid into `evaluate`.

randycoulman committed:
🐛 Detect circular prerequisite flag dependencies

randycoulman committed:
✨ Check for type mismatches

randycoulman committed:
♻️ Rename config modules

I recently discovered that the config module names had been changed in the .Net SDK since the version that was included in the Trello card, so renamed here to match the latest names.

randycoulman committed:
✨ Add inline salt/segment fields

Not yet populated; just getting them in place.

randycoulman committed:
✨ Inline salt and segments after loading configs

randycoulman committed:
✨ Use inline salt and segments

randycoulman committed:
✅ Add override tests for prerequisite flags

randycoulman committed:
✅ Add override tests for salt/segments

randycoulman committed:
♻️ Extract a base test case template for common functionality

Move `fixture_file` and `start_config_cat` to the new `Case` module.

randycoulman committed:
🦺 Validate SDK key format

randycoulman committed:
📝 Add more documentation

Adds documentation to `EvaluationDetails` and `User` to provide more information in the standard Elixir docs.

randycoulman committed:
🧪 Add tests involving evaluation log entries

To make these work, I had to add a `config/config.exs` file to configure the logger in the test environment a bit differently.

I copied all of the evaluation log test files over from the Python SDK, but modified the expected log level strings to match Elixir conventions. I may choose to do some post-processing of either expected or actual log entries instead in a later commit.

randycoulman committed:
✨ Begin updating evaluation logging

randycoulman committed:
✨ Implement String.Chars for User

randycoulman committed:
✨ Continue adding evaluation logging

randycoulman committed:
✨ Logging for prerequiste flag condition evaluation

randycoulman committed:
✨ Add logging for segment condition evaluation

randycoulman committed:
✨ Log user condition evaluation

randycoulman committed:
✅ Only log final result at root

randycoulman committed:
✨ Format comparison values consistently

randycoulman committed:
✨ Log when converting non-string comparison value to string

In order to get the necessary information to log, we now pass a `ComparisonContext` to `UserComparator.compare`. This context includes the condition and key as well as the context_salt and salt.

randycoulman committed:
✨ Log user value type mismatches

randycoulman committed:
✨ Only log missing/invalid user warnings once

Extract evaluation warnings to `EvaluationWarnings`, which starts an Agent to keep track of whether or not the warning has been logged already.

This seemed better than trying to pass back a modified `Context` struct from every function in the call chain to one of the warning functions.

randycoulman committed:
✅ Fix minor spacing issues in logs

randycoulman committed:
♻️ Use default_variation_id from context

randycoulman committed:
✅ Allow for unspecified user serialization order

Elixir does not guarantee the order of maps, so JSON-encoded users in the logs don't necessarily match the expected output.

We now parse the logs, extracting and decoding the logged user. We then compare the rest of the log as well as the decoded users.

randycoulman committed:
✅ Raise errors on missing values

randycoulman committed:
✅ Fix last remaining skipped test

randycoulman committed:
✨ Warn when default value doesn't match setting's type

randycoulman committed:
✅ Run more tests synchronously

More test files are using the same SDK keys which results in test failures when two tests with the same key try to start their own ConfigCat instances. Since we only allow a single instance per SDK key, this fails.

randycoulman committed:
♻️ Encapsulate config format (#114)

* ♻️ Rename settings -> feature flags

This is a start at encapsulating the format of the Config map.

* ♻️ Introduce Config.Preferences module

Refactor existing use of preferences to use new module.

* ♻️ Encapsulate EvaluationFormula and rules

* ♻️ Move entry matching to config modules

randycoulman committed:
♻️ Return entire config (#117)

Always return the entire config, not just the feature flags. This will allow access to preferences and segments in the v6 config where needed.

randycoulman committed:
✅ Use capture_log instead of with_log

with_log wasn't introduced until Elixir 1.13, but we still run CI on 1.12.

randycoulman committed:
🔧 Use older Logger config in Elixir < 1.15

randycoulman committed:
✅ Use proper log level for different Elixir versions

< 1.15: `warn`
>= 1.15: `warning`

randycoulman committed:
🐛 Replace all instances of warning

Use multiline mode on the regex.

randycoulman committed:
🐛 Use warning in 1.13 and 1.14

I mistakenly believed the change came in 1.15, but apparently it came in 1.13.

kp-cat commented on lib/config_cat/config.ex:
```diff
@@ -2,19 +2,27 @@ defmodule ConfigCat.Config do
 
-  @typedoc "A collection of configuration settings and preferences."
-  @type t :: map()
+  @typedoc "A collection of feature flags and preferences."
```
```suggestion
  @typedoc "A collection of configuration settings and preferences."
```
I would keep the original here

randycoulman committed:
📝 Minor doc fix

Co-authored-by: Peter Adam Korodi <52385411+kp-cat@users.noreply.github.com>

kp-cat commented on test/evaluation_log_test.exs:
```diff
@@ -0,0 +1,141 @@
+  end
+
+  test "comparators validation" do
+    # self.maxDiff = None
```
Is this line needed here?

randycoulman replied:
Nope, left over from when I copied in the python code. I wasn't sure if I'd need something like it and then forgot to delete it.

kp-cat commented on test/flag_override_test.exs:
```diff
@@ -151,12 +158,102 @@ defmodule ConfigCat.FlagOverrideTest do
+    end
   end
 
+  # defp start_config_cat(sdk_key, options) do
+  #   name = String.to_atom(UUID.uuid4())
+
+  #   default_options = [
+  #     fetch_policy: CachePolicy.lazy(cache_refresh_interval_seconds: 300),
+  #     name: name,
+  #     sdk_key: sdk_key
+  #   ]
+
+  #   start_supervised!({ConfigCat, Keyword.merge(default_options, options)})
+  #   {:ok, name}
+  # end
```
Can you please remove the unused code here?

randycoulman committed:
🔥 Remove commented code

randycoulman committed:
✅ Add test to show that cache time is respected

Adds a test to show that the lazy cache's TTL is checked against the current cache entry's time before refetching.

adams85 commented on lib/config_cat/rollout.ex:
```diff
@@ -98,165 +237,338 @@ defmodule ConfigCat.Rollout do
+
+      {:ok, comparison_attribute} ->
+        case User.get_attribute(user, comparison_attribute) do
+          nil ->
```
Sadly, my knowledge of Elixir is pretty limited but I assume this pattern won't match empty strings.

However, empty strings should also be treated as a missing attribute value. See https://github.com/configcat/common-js/blob/1ec2efe7bcdc4245c83543d00bbfc966dbd76cd2/src/RolloutEvaluator.ts#L323 (This is a backward compatibility thing.)

randycoulman replied:
Fixed in b10323d

adams85 commented on lib/config_cat/config/user_comparator.ex:
```diff
@@ -0,0 +1,525 @@
+
+  defp to_float(value) when is_binary(value) do
+    value
+    |> String.replace(",", ".")
```
As per the [specification](https://docs.google.com/document/d/1-Vi0YnX57P3gW1-_K_kaR1uZECQowajGJtbGwYvb3a8/#heading=h.lxt05eeqv4rp), leading/trailing whitespace in the user attribute value should be ignored in the case of number and date comparators.

According to my tests, `Float.parse()` doesn't allow leading/trailing whitespace, so we need to trim the string.

```suggestion
    |> String.trim()
    |> String.replace(",", ".")
```

randycoulman replied:
Fixed in 035cacb

adams85 commented on lib/config_cat/rollout.ex:
```diff
@@ -98,165 +237,338 @@ defmodule ConfigCat.Rollout do
 
-    result
+          result = evaluate_user_condition(condition, name, context)
+          # Faking multiple conditions; may want to use actual condition count
+          # eventually. Keeping it this way to match Python SDK for now.
+          EvaluationLogger.log_evaluating_user_condition_result(logger, result, 2)
```
W.r.t. segments, there are two special cases regarding logging:
1. When user object is missing, we omit the details of the segment evaluation.
2. When user object is missing and the targeting rule contains no more conditions besides the segment condition (that is, condition count == 1), we put the THEN part in the same line to make the log more concise (and to be consistent with the way user conditions are logged). Otherwise, we put THEN in a separate line.

So, condition count plays a role in positioning the THEN part. This is the correct logic: https://github.com/configcat/common-js/blob/1ec2efe7bcdc4245c83543d00bbfc966dbd76cd2/src/RolloutEvaluator.ts#L279

randycoulman replied:
The code referenced above is about how we log the result of evaluating a user condition inside of a segment condition, not about how the segment condition itself is logged. The specific log output is either `=> true` or `=> false, skipping the remaining AND conditions`.

In the Python code that I used as a reference, we have the following cases:

1. When evaluating a top-level user condition, we only print that output if there are >= 1 condition; otherwise nothing is printed.
2. When evaluating a user condition within a segment condition, we print that message regardless of how many conditions there are.

By passing a `2` here, I can reuse the code for printing the message and force it to always print the message even if there is only one condition. That feels a bit ugly, which is why I added the comment.

If that logic is incorrect, let me know how I should fix it.

adams85 replied:
> The code referenced above is about how we log the result of evaluating a user condition inside of a segment condition, not about how the segment condition itself is logged

Sorry, this was an oversight on my part.

As for the segment condition special cases:

> When user object is missing and the targeting rule contains no more conditions besides the segment condition (that is, condition count == 1), we put the THEN part in the same line to make the log more concise (and to be consistent with the way user conditions are logged). Otherwise, we put THEN in a separate line.

This is checked by https://github.com/configcat/elixir-sdk/blob/cc9361b27067fcba92c3b3f4fde5a28267684a83/test/support/fixtures/evaluation/segment/segment_no_user.txt

> When user object is missing, we omit the details of the segment evaluation.

Unfortunately, this case has not been tested so far. But now I specified a test case for this too: https://github.com/configcat/common-js/pull/102/commits/637c5ed8ad9539d074bb656513b4df1a36cf9051

Could you add this to the Elixir SDK's test suite? (And, @kp-cat, also to the Python SDK's?)

randycoulman replied:
Done in 00b78c3. The new test is passing, but some others are not yet.

adams85 commented on lib/config_cat/rollout.ex:
```diff
@@ -98,165 +237,338 @@ defmodule ConfigCat.Rollout do
+
+        unless setting_type == comparison_value_type do
+          raise EvaluationError,
+                "Type mismatch between comparison value type #{SettingType.to_elixir_type(comparison_value_type)} and type #{SettingType.to_elixir_type(setting_type)} of prerequisite flag '#{prerequisite_key}'"
```
Could we include the comparison value in the message like it's done here: https://github.com/configcat/common-js/blob/1ec2efe7bcdc4245c83543d00bbfc966dbd76cd2/src/RolloutEvaluator.ts#L626 ?

If you find it useful, I'm ok with including types as well but then we'd rather include the inferred ConfigCat setting types instead of the Elixir types. I'm thinking of a message like this:

`Type mismatch between comparison value '{COMPARISON_VALUE}' (inferred type: {COMPARISON_VALUE_SETTING_TYPE}) and prerequisite flag '{PREREQ_FLAG_KEY}' (type: {PREREQ_FLAG_SETTING_TYPE}).`

randycoulman replied:
I was following what the Python SDK here does, so you might want to make the same change there. I'll leave the types out to match the JS version.

randycoulman replied:
Addressed in c22c00c

adams85 commented on lib/config_cat/rollout.ex:
```diff
@@ -98,165 +237,338 @@ defmodule ConfigCat.Rollout do
+  defp evaluate_percentage_options(percentage_options, %Context{} = context) do
+    case extract_user_key(context) do
+      {:ok, user_key} ->
+        hash_val = hash_user(user_key, context.key)
```
`user_key` can also be of other types than string. So, before hashing, we need to make sure that the attribute value is converted to the "canonical" string representation: https://github.com/configcat/common-js/blob/1ec2efe7bcdc4245c83543d00bbfc966dbd76cd2/src/RolloutEvaluator.ts#L220C1-L220C1 As far as I can see, this conversion (`user_value_to_string`) is missing currently. 

BTW, this is important because the hash values should by "sticky" on all platforms. (We should have specified tests for ensuring this but, sadly, forgot to do so...)

randycoulman replied:
Addressed in 122bce1

adams85 commented on lib/config_cat/config/setting.ex:
```diff
@@ -0,0 +1,102 @@
+    if variation_id(setting) == variation_id do
+      value(setting)
+    else
+      case targeting_rule_variation_value(setting, variation_id) do
```
What about the case when the THEN part of the targeting rule is a list of percentage options? As far as I can tell, this case is not covered. See also: https://github.com/configcat/common-js/blob/1ec2efe7bcdc4245c83543d00bbfc966dbd76cd2/src/ConfigCatClient.ts#L505-L516

randycoulman replied:
Ah, yep, I forgot to add that part when I added support for nested percentage options. Nice catch!

randycoulman replied:
Fixed in 5da8c65

adams85 commented on lib/config_cat/client.ex:
```diff
@@ -197,29 +203,26 @@ defmodule ConfigCat.Client do
-      {key, Map.get(value, Constants.value())}
+    case Setting.variation_value(setting, variation_id) do
+      nil -> nil
+      value -> {key, value}
```
We should check that `value` is an allowed setting value. (Unfortunately, using flag overrides users can pass whatever garbage they want to the evaluation logic.) See also: https://github.com/configcat/common-js/blob/1ec2efe7bcdc4245c83543d00bbfc966dbd76cd2/src/ConfigCatClient.ts#L499

randycoulman replied:
I don't see this check anywhere in the Python SDK. Should it be there as well?

Looking through the JS code, it seems like this check (or at least `isAllowedValue`) appears in many places, not just here. Should it be added elsewhere as well?

randycoulman replied:
Addressed in 5c9b4f7

adams85 replied:
> I don't see this check anywhere in the Python SDK. Should it be there as well?

Yep, looks like Python SDK misses this check. /cc @kp-cat 

> Looking through the JS code, it seems like this check (or at least isAllowedValue) appears in many places, not just here. Should it be added elsewhere as well?

Yes, we should guarantee that when the evaluation is successful, the SDK always returns a valid feature flag value (bool, string, int or double).

adams85 replied:
Thinking it all through again, I can see the following problematic cases that can be caused by invalid flag overrides:

1. Setting type is missing from the config json or it is invalid. E.g.
    ```json
    {
      "t": 6, // unsupported type
      "v": { /*...*/ }
    }
    ```

    This is already handled here: https://github.com/configcat/elixir-sdk/blob/00289e56e29eed57e3c76afa67cc321a2494fc25/lib/config_cat/config/setting_value.ex#L24-L25

3. Setting type is valid but the return value is missing. E.g.
    ```json
    {
      "t": 0, // boolean
      "v": {
        "s": true // the wrong property is set ("b" should be set)
      }
    }
    ```

    This is handled here: https://github.com/configcat/elixir-sdk/blob/00289e56e29eed57e3c76afa67cc321a2494fc25/lib/config_cat/config/setting_value.ex#L29-L31

4. Setting type is valid, the return value is present but it is invalid. E.g.
    ```json
    {
      "t": 0, // boolean
      "v": {
        "b": "true" // this is not a boolean value!
      }
    }
    ```

   This case is not handled currently, isn't it? If my understanding is correct, "b" can be whatever garbage, yet will be returned to the caller.
   In such cases the evaluation will be seemingly successful but return a nonsensical value. I think we should avoid this.
   So we need to check that the type of the return value and the setting type actually match. We should do this in every case where the evaluation is successful and returns a value. (Please forget about the JS SDK's code. It wasn't a good idea to cite that because the JS SDK doesn't support complex flag overrides, so it doesn't need to handle these edge cases.)

randycoulman replied:
Addressed in 43d9d89.

adams85 commented on lib/config_cat/client.ex:
```diff
@@ -249,23 +259,48 @@ defmodule ConfigCat.Client do
+      true ->
+        ConfigCatLogger.warning(
+          "The type of a setting does not match the type of the specified default value (#{default_value}). " <>
+            "Setting's type was #{value_type} but the default value's type was #{default_type}. " <>
```
As `value_type`, please use the ConfigCat setting type instead of the Elixir type (for consistency with other SDKs).

randycoulman replied:
Again, the Python SDK uses the Python type in both places. The JS SDK uses the ConfigCat setting type for `value_type`, but the JS type for `default_type`. I'll change Elixir to use the ConfigCat type for both, but let me know which you prefer.

randycoulman replied:
Addressed in 4130cee

adams85 replied:
> The JS SDK uses the ConfigCat setting type for value_type, but the JS type for default_type

I think this is the correct approach because the default value is not required to be of a valid setting type.

@kp-cat, would be nice to fix this minor inconsistency in Python SDK as well.

randycoulman replied:
Elixir doesn't have a clean way to determine the type of an arbitrary value, so it would be difficult to implement these. There are [workarounds](https://stackoverflow.com/a/40820903), but they're non-standard and potentially fragile.

Is it OK to leave the Elixir implementation as-is?

adams85 replied:
Ok, this is not something critical, so if there's no clean way to do this, let's leave it as-is. 

I talked to @kp-cat about this, and he said that he wouldn't want to change this in Python either.

adams85 commented on lib/config_cat/config/user_comparator.ex:
```diff
@@ -0,0 +1,525 @@
+
+  defp to_string_list(value) when is_binary(value) do
+    case Jason.decode(value) do
+      {:ok, decoded} when is_list(decoded) -> {:ok, decoded}
```
Unfortunately, checking that the parsed object is a list doesn't guarantee that it does contain string values only.

The Python SDK seems to miss this check too, while e.g. the JS SDK not:
* https://github.com/configcat/common-js/blob/1ec2efe7bcdc4245c83543d00bbfc966dbd76cd2/src/RolloutEvaluator.ts#L786
* https://github.com/configcat/common-js/blob/1ec2efe7bcdc4245c83543d00bbfc966dbd76cd2/src/RolloutEvaluator.ts#L716

So, I'm not completely sure about this. However, checking the type of the items would result in a more failsafe logic, so it seems useful to me to add this check (and in Python, as well). 

@kp-cat What do you think?

kp-cat replied:
I think we can do the same check here 👍 
(This is the python version: https://github.com/configcat/python-sdk/blob/master/configcatclient/rolloutevaluator.py#L186)

randycoulman replied:
I wondered about that, but forgot to ask.

randycoulman replied:
Done in 4f3ffd8

adams85 commented on test/rollout_test.exs:
```diff
@@ -1,70 +1,125 @@
   @variation_test_type "variation_test"
 
   test "basic rule evaluation" do
+    # https://app.configcat.com/v2/e7a75611-4256-49a5-9320-ce158755e3ba/08d5a03c-feb7-af1e-a1fa-40b3329f8bed/08dbc4dc-1927-4d6b-8fb9-b1472564e2d3/244cf8b0-f604-11e8-b543-f23c917f9d8d
```
Could you add rollout matrix tests for the old SDK keys as well? See: https://github.com/configcat/common-js/blob/1ec2efe7bcdc4245c83543d00bbfc966dbd76cd2/test/MatrixTests.ts#L19-L40

randycoulman replied:
The Python SDK does not have these tests. You might want to add them there as well.

randycoulman replied:
Added in 7e8e2f2

adams85 replied:
> The Python SDK does not have these tests. You might want to add them there as well.

Hmm, that's weird. As far as I can remember, when I was reviewing the Python PR, it included those tests... But now it doesn't indeed. @kp-cat Could you also re-add these tests?

adams85 commented on test/integration_test.exs:
```diff
@@ -2,39 +2,78 @@ defmodule ConfigCat.IntegrationTest do
-  @tag capture_log: true
-  test "raises error when starting another instance with the same SDK key" do
-    {:ok, _} = start_config_cat(@sdk_key, name: :original)
+    for sdk_key <- [
+          "key",
+          "configcat-proxy/key",
+          "1234567890abcdefghijkl01234567890abcdefghijkl",
+          "configcat-sdk-2/1234567890abcdefghijkl/1234567890abcdefghijkl",
+          "configcat/1234567890abcdefghijkl/1234567890abcdefghijkl"
+        ] do
```
Could you also add these cases: https://github.com/configcat/common-js/blob/1ec2efe7bcdc4245c83543d00bbfc966dbd76cd2/test/ConfigCatClientTests.ts#L25-L41 ?

randycoulman replied:
Again, the Python SDK doesn't have all of these tests. You might want to add them.

randycoulman replied:
Added in bfa583e

adams85 replied:
> Again, the Python SDK doesn't have all of these tests. You might want to add them.

Probably this happened because the Python tests had been specified earlier and we forgot to update them.

@kp-cat However, I think the JS SDK's test suite is more comprehensive, so it'd be nice to have these tests in Python as well.

adams85 reviewed (changes requested):
Hi @randycoulman,

I'm on the ConfigCat team and was asked to do a review on your PR.

I found only minor issues, mostly around edge cases, so I can certainly say that this is a very solid implementation! Even more so, because I know that porting procedural code to functional is not an easy task. So, very well done, sir! 👍

Apart from the comments below, I'd like to ask you to implement these recent changes:
* Please update the message of log event 1103 as shown here: https://github.com/configcat/common-js/pull/100/commits/50aeed44c1fde5ec4155bad5e7f9b84830a5b595
* Please implement this additional test: https://github.com/configcat/js-sdk/blob/master/test/SpecialCharacterTests.ts

Thank you.

randycoulman committed:
🐛 Treat empty string as missing attribute value

randycoulman committed:
🔊 Change wording of type mismatch errors

randycoulman committed:
🐛 Trim strings before converting to floats

randycoulman committed:
🐛 Convert user attribute values to strings before hashing

randycoulman committed:
✨ Look at nested percentage options for get_key_and_value

randycoulman committed:
✨ Warn if setting value is of an unsupported type

randycoulman committed:
🔊 Use ConfigCat types instead of Elixir types

randycoulman committed:
✨ Check for string contents of JSON list

randycoulman committed:
✅ Add config v1 matrix tests

randycoulman committed:
✅ Add more SDK key validation tests

randycoulman committed:
🔊 Add more details to 1103 error message

randycoulman committed:
✅ Add special characters tests

randycoulman commented:
@adams85 New features are implemented in 4ca885b (updated log message) and d07c796 (special characters tests).

randycoulman committed:
🏷️ Fix typespec on user_value_to_string

randycoulman committed:
✅ Fix case-sensitivity issue in matrix filename

adams85 commented on lib/config_cat/config/user_comparator.ex:
```diff
@@ -0,0 +1,539 @@
+
+  def user_value_to_string(value) when is_list(value) do
+    with {:ok, list} <- to_string_list(value) do
+      {:ok, to_string(list)}
```
We want the `user_value_to_string` and `to_xxx` functions to be symmetric, i.e. `user_value_to_string` should produce a value, which is accepted by `to_xxx` and vice versa.

However, according to my tests `to_string(["Item1", "Item2", "Item3"])` results in `Item1Item2Item3`, which is not accepted by `to_string_list`. Instead, we need to JSON encode the list here (`["Item1", "Item2", "Item3"]"`).

See also: https://github.com/configcat/common-js/blob/68e23064e70363614c6f8d8bab9619a67fc32646/src/RolloutEvaluator.ts#L731

@kp-cat Unfortunately, Python SDK seems to have this issue as well.

randycoulman replied:
Addressed in 7ed2557.

adams85 commented on lib/config_cat/config/user_comparator.ex:
```diff
@@ -0,0 +1,539 @@
+
+  defp to_float(_value), do: {:error, :invalid_float}
+
+  defp to_string_list(value) when is_list(value), do: {:ok, value}
```
As far as I can tell, this doesn't prevent list items other then strings, right? I think we need the same check here as below.

See also: https://github.com/configcat/common-js/blob/68e23064e70363614c6f8d8bab9619a67fc32646/src/RolloutEvaluator.ts#L781-L788

randycoulman replied:
Addressed in 00289e5.

adams85 reviewed (changes requested):
In the meantime we specified a few additional tests: https://github.com/configcat/common-js/pull/102

Could you add these to the Elixir SDK's test suite?

Oh, and a minor observation about the sample projects: I had to apply [this workaround](https://github.com/elixir-lang/elixir/issues/12681#issuecomment-1598312912) to be able to run the samples. (I'm on Win10). Could you maybe do something about this if it can be avoided?

randycoulman committed:
⬆️ Update deps in sample apps

Brings them into line with the main app and should also fix an issue with `ssl_verify_fun` 1.1.6.

randycoulman committed:
✅ Add new cache key tests

randycoulman committed:
✅ Add tests for value trimming

One bug fix included: Python and JS allow whitespace around numeric values whereas Elixir does not, which broke the starts_with/ends_with hashed tests while attempting to parse the length string. We add trimming to just the length string to avoid this issue and match Python/JS functionality.

randycoulman committed:
🚧 WIP: add new log tests

Currently failing due to change in logic in JS version; will need to investigate further.

randycoulman committed:
🐛 Fix conversion of string list -> string

randycoulman committed:
✨ Ensure list contains all strings

randycoulman commented:
The fix for the sample projects is in 1005612. I just had to update the dependencies to match what was in the main project.

New tests are in 767ec96, 4c5acac, and 00b78c3. They are not quite passing yet, because the logging logic is different between the Python and JS versions and I ran out of time for tonight before I could resolve the remaining differences. I'll take another look at it tomorrow night.

adams85 commented on lib/config_cat/config/user_comparator.ex:
```diff
@@ -0,0 +1,545 @@
+
+  def user_value_to_string(value) when is_list(value) do
+    with {:ok, list} <- to_string_list(value) do
+      {:ok, inspect(list)}
```
Unfortunately, a further issue has been uncovered regarding this conversion: even that matters how the JSON string is formatted.

After discussing this, we had to make an addition to the specification. This explains the why: https://docs.google.com/document/d/1-Vi0YnX57P3gW1-_K_kaR1uZECQowajGJtbGwYvb3a8/edit#heading=h.6ajheae6qud8

Also, I specified a new test to ensure the platform-independent behavior across our SDKs. You can find it here: https://github.com/configcat/common-js/pull/102/commits/f6290509b14ce17740447d5b056c9e972d676a8d

Please make the necessary adjustment and implement this test.

And thank you for your patience, expectedly this will be my last request to you in this PR.

randycoulman replied:
Done in b1194b3. This was quite tricky, as Elixir's float formatting is different than JavaScripts and I had to find a way to make them match.

randycoulman committed:
✅ Add (and pass) new string conversion tests

Making Elixir's standard float formatting match JavaScript's was somewhat tricky.

randycoulman committed:
🥅 Raise an error when a setting value doesn't match its key

e.g. `%{"b" => "true"}` is invalid.

randycoulman commented:
I've addressed the two new requests, but am still working on fixing the logging tests. I'll continue tomorrow.

randycoulman committed:
♻️ Minor cleanups to logging code

- Reverse names of `log_evaluating_condition_result` and `log_evaluating_condition_final_result` to better reflect their usage

- Move condition_count check to caller for `log_evaluating_condition_result`

- Move `log_evaluating_targeting_rules` to its proper alphabetical position

♻️ Refactor log_evaluating_condition_final_result

Pass a `newline?` param instead of `condition_count` to prepare for a better calculation of `newline?`.  Involved moving some increase_/decrease_indent calls around.

✨ Pass newline_before_then? flag

Based on the final evaluated condition, determine whether or not there should be a newline before the `THEN` in the logs.

This was a bit tricky to properly accumulate and pass around.

randycoulman committed:
✅ Fix remaining log test

Compute and return a `newline_before_then?` flag and pass it to the
logger.

randycoulman commented:
@adams85 I now have all of the tests passing. I had to add the `newline_before_then` logic from the JS version to get that to work, and that ended up being somewhat ugly. There's not really any way around that, though.

I believe I've now addressed all of your feedback, so I think this should be good to merge. Let me know if you find anything else.

adams85 commented:
Thanks for taking the trouble. 🙏 As for me, the PR is complete.
